File: 130-surrounded-regions/130-surrounded-regions.go
func solve(board [][]byte)  {
    m, n := len(board), len(board[0])
    
    var dfs func(r, c int)
    dfs = func(r, c int) {
        if r < 0 || c < 0 || r >= m || c >= n || board[r][c] != 'O' {
            return
        }
        board[r][c] = '#'
        dfs(r - 1, c)
        dfs(r + 1, c)
        dfs(r, c - 1)
        dfs(r, c + 1)
    }
    
    // Mark the first and last col
    for i := 0; i < m; i++ {
        if board[i][0] == 'O' {
            dfs(i, 0)
        }
        if board[i][n - 1] == 'O' {
            dfs(i, n - 1)
        }
    }
    
    // Mark the first and last row
    for j := 0; j < n; j++ {
        if board[0][j] == 'O' {
            dfs(0, j)
        }
        if board[m - 1][j] == 'O' {
            dfs(m - 1, j)
        }
    }
    
    // replace any remaining Os, and restore marked #s
    for i := 0; i < m; i++ {
        for j := 0; j < n; j++ {
            if board[i][j] == '#' {
                board[i][j] = 'O'
            } else {
                board[i][j] = 'X'
            }
        }
    }
}

File: 130-surrounded-regions/130-surrounded-regions.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """Do not return anything, modify board in-place instead."""
        m, n = len(board), len(board[0])
        offsets = (
            (-1, 0),  # up
            (1, 0),   # down
            (0, -1),  # left
            (0, 1)    # right
        )
        
        def dfs(r: int, c:int):
            if not (0 <= r < m) or not (0 <= c < n) or board[r][c] != "O":
                return
            board[r][c] = "#"
            for dr, dc in offsets:
                dfs(r + dr, c + dc)
        
        # Mark the first and last columns
        for i in range(m):
            if board[i][0] == "O":
                dfs(i, 0)
            if board[i][n - 1] == "O":
                dfs(i, n - 1)
                
        # Mark the first and last rows
        for j in range(n):
            if board[0][j] == "O":
                dfs(0, j)
            if board[m - 1][j] == "O":
                dfs(m - 1, j)
        
        # Replace remaining Os in-place, and restore marked #s
        board[:] = [["XO"[c == "#"] for c in row] for row in board]